[PORT] Removes overlay queuing, saves 6/7 seconds of initialize. Lightly modifies stat tracking macros (+some sprite error fix)

EvilDragonfiend opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
* https://github.com/tgstation/tgstation/pull/69696

>So we have this overlay queuing system right? It's built with the assumption that the "add to overlay list" operation is real expensive, and is thus useful to queue removals or additions.
>
>It turns out that it just isn't, at least during init. In my testing the operation of queuing took LONGER then the actual overlay add/remove did.
>
>That's ignoring the cost of the subsystem's work.
>
>I've also modified part of the stat tracking macro, since it took a good bit of cpu time, and didn't seem to well, do anything. So far as I can tell it always evaluates to 1

This PR does likely bug fix and corrects a weird behaviour
There is some removal of zmimic code but it's fully okay and I checked daedalus overylay subsystem follows this TG version.

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
bug fix

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding.
-->

## Testing Photographs and Procedure
<!-- Include any screenshots/videos/debugging steps of the modified code functioning successfully, ideally including edge cases. -->
![image](https://github.com/BeeStation/BeeStation-Hornet/assets/87972842/62140a5d-7331-47f3-aedb-f8fe1a4e1f1e)

my character is hiding in a pot

![image](https://github.com/BeeStation/BeeStation-Hornet/assets/87972842/823e8a9e-1d8b-46d1-b847-173652d8f668)

fire alarm respects alert level

![image](https://github.com/BeeStation/BeeStation-Hornet/assets/87972842/3e3b4702-67d9-42eb-9dcf-98bdf70d7b94)

cult halo okay

![image](https://github.com/BeeStation/BeeStation-Hornet/assets/87972842/6dedf6cf-d673-4942-91e8-0ea940c56842)

mob overlays okay

## Changelog
:cl:
code: mildly cleaned up flags_1 defines
code: ported a refactor of subsystem overlay from TG. Overlay list is now trustful.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

EvilDragonfiend committed:
Overlay port

EvilDragonfiend committed:
decal fix

EvilDragonfiend committed:
correct define REFERENCE_TRACKING_FAST

Tsar-Salat commented on code/__DEFINES/flags.dm:
```diff
@@ -27,18 +27,24 @@ GLOBAL_LIST_INIT(bitflags, list(1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 204
+#define HTML_USE_INITAL_ICON_1		(1<<10) 			//! Should we use the initial icon for display? Mostly used by overlay only objects
+
+//turf-only flags
+#define NOJAUNT_1					(1<<11)
```
Guh this flags_1 stuff is just getting more bloated.

Instead of making the turf-only flags_1 conform to the numeric numbering, why not split them off into turf_flags like tg did? 

Seems a lot less ugly than having so many bitflags

Tsar-Salat replied:
(tell me off if this is too big a change ofc, it would be quite a few lines)

EvilDragonfiend replied:
It's relocation. It's already how it's been already. I am not even making it bloated.

Tsar-Salat commented on code/__DEFINES/flags.dm:
```diff
@@ -27,18 +27,24 @@ GLOBAL_LIST_INIT(bitflags, list(1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 204
-#define PREVENT_CONTENTS_EXPLOSION_1 (1<<16)
-#define UNPAINTABLE_1 				(1<<17)
-#define HTML_USE_INITAL_ICON_1		(1<<18) 			//! Should we use the initial icon for display? Mostly used by overlay only objects
+#define CONDUCT_1					(1<<0)		//!  conducts electricity (iron etc.)
```
If you're editing the numbers, can you convert the comments to autodoc? Should make it more legible in github and you can see the comments through the lang extension

Tsar-Salat replied:
Like so:
```
//  conducts electricity (iron etc.)
#define CONDUCT_1 (1<<0)
//  For machines and structures that should not break into parts, eg, holodeck stuff
#define NODECONSTRUCT_1 (1<<1)
//  item has priority to check when entering or leaving
#define ON_BORDER_1 (1<<2)
```

EvilDragonfiend replied:
sure

Tsar-Salat reviewed (changes requested):
Nitpicks

EvilDragonfiend committed:
fixes unit test warnings

EvilDragonfiend committed:
additional icon state fixes

EvilDragonfiend committed:
edits define comments

EvilDragonfiend committed:
gun sprite snowflake

EvilDragonfiend committed:
Fixes sprites v2

EvilDragonfiend committed:
unit test snowflake

EvilDragonfiend committed:
fix inverted condition

EvilDragonfiend commented on code/controllers/subsystem/overlays.dm:
```diff
@@ -95,167 +38,83 @@ SUBSYSTEM_DEF(overlays)
-/atom/movable/build_appearance_list(old_overlays)
-	var/static/image/appearance_bro = new()
-	var/list/new_overlays = list()
+#ifdef UNIT_TESTS
+/atom/var/skip_sprite_error
+/obj/item/gun/skip_sprite_error = TRUE
+// I hate this but gun sprites are not implemented properly, and unit test blames this
+// We'll need to clean up gun sprites someday, but not right now, it's a mess
+#endif
```
~~Error details can be found in:~~
* ~~https://github.com/BeeStation/BeeStation-Hornet/actions/runs/8277897976/job/22649152337~~

Well, it's deleted now because it's too old.
Removing this define will show which wrong icon status we have ingame, especially for guns.

EvilDragonfiend committed:
Lemon guy said it's okay to remove

EvilDragonfiend committed:
Merge branch 'master' into overlayport-pt1

EvilDragonfiend committed:
Merge branch 'master' into overlayport-pt1

EvilDragonfiend committed:
Merge branch 'master' into overlayport-pt1

EvilDragonfiend committed:
Merge remote-tracking branch 'BeeStation/master' into overlayport-pt1

EvilDragonfiend committed:
cell fix

EvilDragonfiend committed:
Merge remote-tracking branch 'BeeStation/master' into overlayport-pt1

EvilDragonfiend committed:
'd007' to 'd00-7'

Invalid overlay: Icon object 'icons/obj/dice.dmi' [0xc000b7a] used in 'the d00' /obj/item/dice/d00 is missing icon state d00-7.

PowerfulBacon commented on code/game/objects/effects/decals/decal.dm:
```diff
@@ -45,7 +45,17 @@
+	if(GLOB.running_create_and_destroy)
+		var/turf/T = loc
+		T.RemoveElement(/datum/element/decal, icon, icon_state, dir, null, layer, alpha, color, null, FALSE, null)
+#endif
```
This is pretty horrendous. Can you make it just never apply decals instead of UNIT_TESTS is defined and remove the GLOB.running_create_and_destroy var?

EvilDragonfiend replied:
I doubt if it's a good idea since it's 100% TG port.
This global variable is specifically for when `/datum/unit_test/create_and_destroy` is running but also when decals are being created. When `/datum/unit_test/create_and_destroy` test is over, `GLOB.running_create_and_destroy` becomes FALSE, and we can possibly have more decals in creation after this test, and just `#define UNIT_TEST` will remove those even when we don't want these are removed.

PowerfulBacon reviewed (changes requested):
Has this been tested with multi-z?

EvilDragonfiend committed:
We won't have GLOB.running_create_and_destroy

EvilDragonfiend committed:
ugly extra line

EvilDragonfiend committed:
Merge branch 'master' into overlayport-pt1

EvilDragonfiend committed:
BACON WHY

EvilDragonfiend committed:
cyborg overylay fix

EvilDragonfiend committed:
Merge branch 'master' into overlayport-pt1

EvilDragonfiend committed:
Merge remote-tracking branch 'BeeStation/master' into overlayport-pt1

EvilDragonfiend committed:
Removes unused macro

PowerfulBacon reviewed (changes requested):
Breaks gun overlays

EvilDragonfiend committed:
Merge remote-tracking branch 'BeeStation/master' into overlayport-pt1

EvilDragonfiend committed:
removes magazine check

EvilDragonfiend committed:
fixes gun sprite error

EvilDragonfiend committed:
Merge remote-tracking branch 'BeeStation/master' into overlayport-pt1

EvilDragonfiend committed:
solves atmos console overlay

EvilDragonfiend committed:
Merge remote-tracking branch 'BeeStation/master' into overlayport-pt1

EvilDragonfiend committed:
Merge branch 'master' into overlayport-pt1

EvilDragonfiend committed:
Update bitfields.dm
